crates/mysten-common/src/sync/notify_read.rs

```diff
@@ -116,22 +128,100 @@ impl<K: Eq + Hash + Clone, V: Clone> NotifyRead<K, V> {
+            .map(|(key, _result)| key.clone())
+            .collect();
+        let has_waiting_keys = !waiting_keys.is_empty();
+        let waiting_keys = Arc::new(Mutex::new(waiting_keys));
+
+        // Spawn logging task if there are waiting keys
+        let log_handle = if has_waiting_keys {
+            let waiting_keys_clone = waiting_keys.clone();
+            let start_time = Instant::now();
+            let task_name = task_name.to_string();
+
+            Some(spawn_monitored_task!(async move {
+                // Only start logging after the first interval.
+                let start = Instant::now() + Duration::from_secs(LONG_WAIT_LOG_INTERVAL_SECS);
+                let mut interval =
+                    interval_at(start, Duration::from_secs(LONG_WAIT_LOG_INTERVAL_SECS));
+
+                loop {
+                    interval.tick().await;
+                    let current_waiting = waiting_keys_clone.lock();
+                    if current_waiting.is_empty() {
+                        break;
+                    }
+                    let keys_vec: Vec<_> = current_waiting.iter().cloned().collect();
+                    drop(current_waiting); // Release lock before logging
+
+                    let elapsed_secs = start_time.elapsed().as_secs();
+
+                    warn!(
+                        "[{}] Still waiting for {}s for {} keys: {:?}",
+                        task_name,
+                        elapsed_secs,
+                        keys_vec.len(),
+                        keys_vec
+                    );
+
+                    if task_name == "CheckpointBuilder::resolve_checkpoint_transactions"
+                        && elapsed_secs >= 60
+                    {
+                        debug_fatal!("CheckpointBuilder::resolve_checkpoint_transactions is stuck");
+                    }
+                }
+            }))
+        } else {
+            None
+        };
+
+        let results =
+            results
+                .into_iter()
+                .zip(registrations)
+                .zip(keys.iter())
+                .map(|((a, r), key)| match a {
+                    // Note that Some() clause also drops registration that is already fulfilled
+                    Some(ready) => Either::Left(futures::future::ready(ready)),
+                    None => {
+                        let waiting_keys = waiting_keys.clone();
+                        let key = key.clone();
+                        Either::Right(async move {
+                            let result = r.await;
+                            // Remove this key from the waiting set
+                            waiting_keys.lock().remove(&key);
+                            result
+                        })
+                    }
+                });
+
+        let resolved_values = join_all(results).await;
+
+        // Cancel the logging task
+        if let Some(handle) = log_handle {
+            handle.abort();
+        }
```

Comment:
I think it's possible to never reach this `abort` here if the overall `read` method is cancelled/aborted in earlier wait methods (ex on line 217 `join_all(results).await;`). Assuming I get this correctly, then the task will run forever in the loop? 

Maybe use a `CancelOnDrop` like https://github.com/MystenLabs/sui/blob/9bdb5287f2073978f58e3a908be678389916c9b8/crates/sui-core/src/consensus_adapter.rs#L1203 or a [JoinSet](https://docs.rs/tokio/latest/tokio/task/struct.JoinSet.html) (although JoinSet might be too much here)